### PR TITLE
Fix for trainer estimator metadata propagation

### PR DESCRIPTION
**Pull request by Zruty0**
Added tests for metadata propagation on existing trainers, also fixed SDCA to pass metadata correctly.

### CODE REVIEW

File: src/Microsoft.ML.Data/Training/TrainerEstimatorBase.cs [hunk 1]
```diff
@@ -62,6 +63,20 @@ public TrainerEstimatorBase(IHost host, SchemaShape.Column feature, SchemaShape.
             WeightColumn = weight;
         }
 
+        /// <summary>
+        /// Normal metadata that we produce for score columns.
+        /// </summary>
+        protected static IEnumerable<SchemaShape.Column> MetadataForScoreColumn(bool isNormalized = false)
```
Comment (sfilipi) [inline src/Microsoft.ML.Data/Training/TrainerEstimatorBase.cs]: >isNormalized [](start = 85, length = 12)

will this come from the trainerInfo, or it is always true for the Propability type column?

File: src/Microsoft.ML.StandardLearners/Standard/Online/AveragedPerceptron.cs [hunk 2]
```diff
@@ -59,9 +59,9 @@ public AveragedPerceptronTrainer(IHostEnvironment env, Arguments args)
 
             _outputColumns = new[]
             {
-                new SchemaShape.Column(DefaultColumnNames.Score, SchemaShape.Column.VectorKind.Scalar, NumberType.R4, false),
-                new SchemaShape.Column(DefaultColumnNames.Probability, SchemaShape.Column.VectorKind.Scalar, NumberType.R4, false),
-                new SchemaShape.Column(DefaultColumnNames.PredictedLabel, SchemaShape.Column.VectorKind.Scalar, BoolType.Instance, false)
+                new SchemaShape.Column(DefaultColumnNames.Score, SchemaShape.Column.VectorKind.Scalar, NumberType.R4, false, new SchemaShape(MetadataForScoreColumn())),
+                new SchemaShape.Column(DefaultColumnNames.Probability, SchemaShape.Column.VectorKind.Scalar, NumberType.R4, false, new SchemaShape(MetadataForScoreColumn(true))),
+                new SchemaShape.Column(DefaultColumnNames.PredictedLabel, SchemaShape.Column.VectorKind.Scalar, BoolType.Instance, false, new SchemaShape(MetadataForScoreColumn()))
```
Comment (sfilipi) [inline src/Microsoft.ML.StandardLearners/Standard/Online/AveragedPerceptron.cs]: >Shape.Column(DefaultColumnNames.PredictedLabel [](start = 26, length = 46)

is it intentional to not use also the input label metadata?

Comment (sfilipi) [inline src/Microsoft.ML.StandardLearners/Standard/Online/AveragedPerceptron.cs]: >MetadataForScoreColumn [](start = 154, length = 22)

shoudl this just be called Metadata, since it is used for all output columns?

File: src/Microsoft.ML.Data/Training/TrainerEstimatorBase.cs [hunk 3]
```diff
@@ -62,6 +63,20 @@ public TrainerEstimatorBase(IHost host, SchemaShape.Column feature, SchemaShape.
...
+        {
+            var cols = new List<SchemaShape.Column>();
+            cols.Add(new SchemaShape.Column(MetadataUtils.Kinds.ScoreColumnSetId, SchemaShape.Column.VectorKind.Scalar, NumberType.U4, true));
+            cols.Add(new SchemaShape.Column(MetadataUtils.Kinds.ScoreColumnKind, SchemaShape.Column.VectorKind.Scalar, TextType.Instance, false));
+            cols.Add(new SchemaShape.Column(MetadataUtils.Kinds.ScoreValueKind, SchemaShape.Column.VectorKind.Scalar, TextType.Instance, false));
+            if (isNormalized)
+                cols.Add(new SchemaShape.Column(MetadataUtils.Kinds.IsNormalized, SchemaShape.Column.VectorKind.Scalar, BoolType.Instance, false));
```
Comment (sfilipi) [inline src/Microsoft.ML.Data/Training/TrainerEstimatorBase.cs]: On the Score column generated from the OVA tests, i see yet another column: "Slot Names"
Dees that need to be added to the list of metadata here?

File: src/Microsoft.ML.StandardLearners/Standard/Online/AveragedPerceptron.cs (hunk 2 above)
Comment (sfilipi) [reply src/Microsoft.ML.StandardLearners/Standard/Online/AveragedPerceptron.cs]: Curious: why do you concat to the label metadata of SDCA multiclass, then? To propagate the info about the string - categories conversion? 

---
In reply to: [217769414](https://github.com/dotnet/machinelearning/pull/909#discussion_r217769414) [](ancestors = 217769414,217574282)

File: src/Microsoft.ML.StandardLearners/Standard/LinearClassificationTrainer.cs [hunk 4]
```diff
@@ -21,6 +21,7 @@
 using Microsoft.ML.Runtime.Learners;
 using Microsoft.ML.Runtime.Numeric;
 using Microsoft.ML.Runtime.Training;
+using System.Collections.Generic;
```
Comment (artidoro) [inline src/Microsoft.ML.StandardLearners/Standard/LinearClassificationTrainer.cs]: Would be good to sort these.

File: src/Microsoft.ML.StandardLearners/Standard/Online/AveragedPerceptron.cs (hunk 2 above)
Comment (Zruty0) [reply src/Microsoft.ML.StandardLearners/Standard/Online/AveragedPerceptron.cs]: SDCA propagates key value metadata, that's what I'm doing there. And I think all the multiclass learners do (or should do)

---
In reply to: [217781701](https://github.com/dotnet/machinelearning/pull/909#discussion_r217781701) [](ancestors = 217781701,217769414,217574282)
